### ISSUE TITLE
Android NTLM: Empty Credentials now throws PlatformNotSupportedException

### ISSUE THREAD

**Issue by dotMorten** [dotnet/runtime#91131]
### Description

If an empty network credential is used [a recent change](https://github.com/dotnet/runtime/pull/87930) is now causing the exception `System.PlatformNotSupportedException: 'NTLM authentication is not possible with default credentials on this platform.'` to be thrown rather than treating the empty credential as null / not set.
This is a regression from .NET 7 behavior.

### Reproduction Steps

On .NET Android, run the folllowing code:
```cs
string url = "https://url.to/my-iwa-secured-server";
var handler = new SocketsHttpHandler();
var cre = handler.Credentials;
handler.Credentials = CredentialCache.DefaultCredentials; // this is empty string username+password
//OR: handler.Credentials = new NetworkCredential();
HttpClient client = new HttpClient(handler);            
var result = await client.SendAsync(new HttpRequestMessage(HttpMethod.Get, url)); // Throws
```

### Expected behavior

Returns 401 like .NET 7 would.

### Actual behavior

```
[mono-rt] [ERROR] FATAL UNHANDLED EXCEPTION: System.PlatformNotSupportedException: NTLM authentication is not possible with default credentials on this platform.
[mono-rt]    at System.Net.NegotiateAuthenticationPal.ManagedNtlmNegotiateAuthenticationPal..ctor(NegotiateAuthenticationClientOptions clientOptions)
[mono-rt]    at System.Net.NegotiateAuthenticationPal.Create(NegotiateAuthenticationClientOptions clientOptions)
[mono-rt]    at System.Net.NegotiateAuthenticationPal.ManagedSpnegoNegotiateAuthenticationPal.CreateMechanismForPackage(String packageName)
[mono-rt]    at System.Net.NegotiateAuthenticationPal.ManagedSpnegoNegotiateAuthenticationPal.CreateSpNegoNegotiateMessage(ReadOnlySpan`1 incomingBlob, NegotiateAuthenticationStatusCode& statusCode)
[mono-rt]    at System.Net.NegotiateAuthenticationPal.ManagedSpnegoNegotiateAuthenticationPal.GetOutgoingBlob(ReadOnlySpan`1 incomingBlob, NegotiateAuthenticationStatusCode& statusCode)
[mono-rt]    at System.Net.Security.NegotiateAuthentication.GetOutgoingBlob(ReadOnlySpan`1 incomingBlob, NegotiateAuthenticationStatusCode& statusCode)
[mono-rt]    at System.Net.Security.NegotiateAuthentication.GetOutgoingBlob(String incomingBlob, NegotiateAuthenticationStatusCode& statusCode)
[mono-rt]    at System.Net.Http.AuthenticationHelper.SendWithNtAuthAsync(HttpRequestMessage request, Uri authUri, Boolean async, ICredentials credentials, Boolean isProxyAuth, HttpConnection connection, HttpConnectionPool connectionPool, CancellationToken cancellationToken)
[mono-rt]    at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
[mono-rt]    at System.Net.Http.AuthenticationHelper.SendWithAuthAsync(HttpRequestMessage request, Uri authUri, Boolean async, ICredentials credentials, Boolean preAuthenticate, Boolean isProxyAuth, Boolean doRequestAuth, HttpConnectionPool pool, CancellationToken cancellationToken)
[mono-rt]    at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
[mono-rt]    at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
[mono-rt]    at IWATest.MainPage.OnCounterClicked(Object sender, EventArgs e) in E:\sources.tmp\IWATest\IWATest\MainPage.xaml.cs:line 25
[mono-rt]    at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
[mono-rt]    at Android.App.SyncContext.<>c__DisplayClass2_0.<Post>b__0() in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.App/SyncContext.cs:line 36
[mono-rt]    at Java.Lang.Thread.RunnableImplementor.Run() in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Thread.cs:line 36
[mono-rt]    at Java.Lang.IRunnableInvoker.n_Run(IntPtr jnienv, IntPtr native__this) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net8.0/android-34/mcw/Java.Lang.IRunnable.cs:line 84
[mono-rt]    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PP_V(_JniMarshal_PP_V callback, IntPtr jnienv, IntPtr klazz) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 22
```

### Regression?

Yes

### Known Workarounds

Check whether the credential is empty before setting, but has effect on already released libraries or cross-platform code.

### Configuration

.NET 8 Preview 7 (Android)

### Other information

/CC @wfurt @filipnavara as discussed

**Comment by filipnavara** [comment]
> Check whether the credential is empty before setting

You cannot check that. The value is a placeholder that's always empty.

---

The question is what is the expected behavior when you specify `CredentialCache.DefaultCredentials`, the server requests NTLM/Negotiate, and the platform doesn't support it.

.NET 6 on Linux/macOS could throw `PlatformNotSupportedException`.
.NET 7 will likely return you `401 Unauthorized` and not throw, on Linux/macOS/Android.
.NET 8 is back to throwing `PlatformNotSupportedException`.

`PlatformNotSupportedException` explicitly tells you what you did wrong, what is not supported, and you can avoid it as a developer.
Returning `401 Unauthorized` is slightly more consistent with other error scenarios, but completely hides the underlying issue from the developer and makes it incredibly difficult to debug.

On API level, the .NET 8 change is that `NegotiateAuthentication` API throws the `PlatformNotSupportedException` exception from constructor instead of returning `Unsupported` from first `GetOutgoingBlob` call. In that case it provides the API consumer with more actionable information.

We can change HttpClient to handle `PlatformNotSupportedException` and pass through the `401 Unauthorized` response.

@karelz @wfurt Thoughts?

**Comment by dotMorten** [comment]
My suggestion would be to special-case empty credentials and treat them as if they weren't set. They are in way the same thing.

**Comment by filipnavara** [comment]
> My suggestion would be to special-case empty credentials and treat them as if they weren't set. They are in way the same thing.

`CredentialCache.DefaultCredentials` is always empty, on all platforms, even the ones where SSO works. It simply acts as a placeholder value.

By that definition `NegotiateAuthentication` API already special cases it. On Windows it translates to telling SSPI to use default credentials, on Linux/macOS it translates to GSSAPI acquiring default credentials (or throwing PNSE for `NTLM`-only case), and on Android/tvOS it throws PNSE unconditionally.

What I am proposing is to swallow the PNSE in `SocketsHttpHandler` and pass on the original HTTP response to the caller.

**Comment by SteveSyfuhs** [comment]
If you're going out of your way to specify what credentials are used, and those credentials aren't supported, it should throw. The platform in question doesn't support default credentials. Having it translate to an application-level error is just weird.

```csharp
handler.Credentials = CredentialCache.DefaultCredentials; // this is empty string username+password
```

Additionally, since it's NTLM, I actually think it should always throw PNSE for `DefaultCredentials` as a matter of security hygiene, and should only support supplied credentials, that way it's signaling to the developer how bad this is.

**Comment by dotMorten** [comment]
> If you're going out of your way to specify what credentials are used, 

On the contrary I'm going out of my way not to write platform specific code. If the platform doesn't support default credentials, why does CredentialCache.DefaultCredentials not return null then?

**Comment by filipnavara** [comment]
> why does CredentialCache.DefaultCredentials not return null then?

Firstly, it was never defined that way. It cannot be changed to return `null` now since it's non-nullable.
Secondly, the system may have default credentials for Kerberos but not for NTLM. There's simply not a way to convey that information in `ICredentials`/`NetworkCredential`, and often getting that information can be quite expensive.

**Comment by dotMorten** [comment]
If I have invalid credentials a 401 seems like the right response, which is what .net7 did

**Comment by filipnavara** [comment]
> If I have invalid credentials a 401 seems like the right response, which is what .net7 did

I see where you are coming from, and I tend to agree to a point. The problem with `HttpClient` is that it doesn't tell you outright that it is not supported. In fact, it doesn't even tell you if you make some HTTP request. It literally only tells you when the server requests the Negotiate/NTLM authentication, and that's not a good experience. That's why I feel a special handling there may be in order. If you specify `CredentialCache.DefaultCredentials` and the server offers you Basic authentication I would expect this to result in "incompatible match" and the `401 Unauthorized` returned to the HttpClient.SendAsync caller.

That said, I believe the current behavior of `NegotiateAuthentication` itself is the correct one. You pass a parameter that is invalid for a given platform, and get an exception right away.

**Comment by karelz** [comment]
@wfurt I am curious what you think. Being in line with .NET 7 behavior makes sense to me.

**Comment by filipnavara** [comment]
> Being in line with .NET 7 behavior makes sense to me.

I agree when it comes to the observable behavior on `HttpClient`/`SocketsHttpHandler`.

I'm not sure I agree when it comes to the observable behavior on `NegotiateAuthentication`. For example,
```
var na = new NegotiateAuthentication(new NegotiateAuthenticationClientOptions { Package = "NTLM" });
na.GetOutgoingBlob("", out var status);
```
This will always fail on any non-Windows platform. Would you expect this to throw `PlatformNotSupportedException("NTLM authentication is not possible with default credentials on this platform.")`, or would you expect it to have no exception and `status == NegotiateAuthenticationStatusCode.Unsupported`? What if this is wrapped inside SQL Client logic and you just try to connect to database (PNSE vs some generic auth exception)?

**Comment by dotMorten** [comment]
Not sure I get the sql client argument. If I talk to a webserver that I can't authenticate with, I expect a 401 response code. Having to deal with certain platforms now throwing in certain cases for specific credentials makes no sense to me and makes it really hard to write libraries that run on multiple platforms.

**Comment by filipnavara** [comment]
> Not sure I get the sql client argument.

It's an argument about non-HTTP scenario. `NegotiateAuthentication` is a building block that's used by HTTP, SMTP, and SQL Server, and `NegotiateStream`, among others.

> If I talk to a webserver that I can't authenticate with, I expect a 401 response code. Having to deal with certain platforms now throwing in certain cases for specific credentials makes no sense to me and makes it really hard to write libraries that run on multiple platforms.

With <= .NET 6 you could actually get `Win32Exception` in the authentication scenarios, and I agree that's pretty unreasonable. That's why I fixed it in .NET 7 to avoid that. 

I would argue that anyone using `CredentialCache.DefaultCredentials` on mobile platform may incorrectly be lead to believe that it can work at all. That's not the case on Android, tvOS, and in most cases on iOS. The standard API way to say that something is not supported on a given platform is the `PlatformNotSupportedException`. That said, in this particular case I find it problematic because it doesn't happen at the time you call `handler.Credentials = CredentialCache.DefaultCredentials` but it happens at some later point and it depends on the server requesting the Negotiate/NTLM authentication. Thus throwing `PlatformNotSupportedException` violates the principle of least surprise.

**Comment by filipnavara** [comment]
The alternative to PNSE in context of `NegotiateAuthentication` is to map to the closest `NegotiateAuthenticationStatus`, which in this case would likely be `UnknownCredentials`.

**Comment by SteveSyfuhs** [comment]
In Windows we'll return SEC_E_NO_CREDENTIALS in these sorts of situations. That is our explicit error to signal CREDUI or bail. GSS is also going to return a similar error. UnknownCredentials is probably the equivalent? But also now you're mixing error behaviors between status codes and exceptions.

The follow up question I have is, one way or another this must be dealt with in the error case otherwise the request just won't work, so what exactly is the OP plan to prompt for creds?

There are genuinely two failure modes: client can't use what it has (because it has nothing or the cred is messed up) and the server accepting can't use what it was given for unknown reasons. The former can provide a TON of information about the failure, and the latter can tell you next to nothing. Filip's point early on was that you'd be trading a ton of useful diagnostic information for something that is nearly useless to debug. That is a silly tradeoff.
________________________________

**Comment by filipnavara** [comment]
> UnknownCredentials is probably the equivalent?

Correct. `SEC_E_NO_CREDENTIALS` would translate to `NegotiateAuthenticationStatus.UnknownCredentials`.

> But also now you're mixing error behaviors between status codes and exceptions.

The question is whether it makes sense to "emulate Windows" here, or throw an exception outright.

In case of the low-level `NegotiateAuthentication` API this boils down to:

1) Throwing `PNSE`:
[+] Accurate message for diagnostics
[-] May need extra handling on the API caller side
[ ] That's what the .NET 6 private API surface did

2) Communicating back with the status code:
[+] Closer to what happens on Windows, so easier API to write against
[-] Bad diagnostics (ie. optional tracing has to be turned on to find out the real error)
[ ] That's closer to what .NET 7 public API shipped with, although unintentionally

**Comment by dotMorten** [comment]
I look at it from a developer point of view. Ie, to me it seems weird that I have to deal with things different:
```
handler.Credential = null; // returns 401
handler.Credential = new NetworkCredential("invalid user", "invalid password"); // returns 401
handler.Credential= new NetworkCredential("", ""); // Throws platform not supported on some platforms because user accepted sign-in dialog before entering anything
handler.Credential= CredentialCache.DefaultCredentials; // Throws platform not supported
```

In _all_ of these cases, I need to do the exact same thing: I need to prompt the user for a valid set of credentials. Just like I do for any other 401 response.

**Comment by wfurt** [comment]
Reading the thread I would probably be inclined to support the idea that the low-level API returns specific error e.g. `NegotiateAuthenticationStatus.UnknownCredentials`. It would be up to the callers to deal with that. That can for example trow for `NegotateStream` to make it clear and it may turn it into "silent" 401 failures for HttpClient. While I understand some of the arguments @SteveSyfuhs, the backward compatibility is quite important for us e.g. just change the behavior because we feel it is better is problematic.

I'm not big fan of throwing/catching/ignoring. I can live with 401 for cases when the password happens to be empty. The explicit use of `CredentialCache.DefaultCredentials` seems more problematic as this is never going to work in certain scenarios. The trick here is that for Kerberos it _may_ work but not for NTLM -> something the developer may not control of as the particular mechanisms are selected during authentication. 

From that prospective I'm probably in favor of #91160

**Comment by filipnavara** [comment]
After writing both versions of the code I am inclined to agree that returning status codes is less disrupting for the current API consumers. That said, we should probably have a documentation on how to enable the detailed tracing. Right now the steps are just scattered around in several issues and random snippets at gist.github.com, which is far from ideal.

**Comment by wfurt** [comment]
I think we should take this one as well for 8.0 @karelz as it can be viewed as 8.0 regression.

**Comment by karelz** [comment]
Reopening to track the backport.

**Comment by dotMorten** [comment]
> .NET 6 on Linux/macOS could throw PlatformNotSupportedException ([1](https://github.com/dotnet/runtime/blob/59a1e88eaa84c1d8d34862fcfc5339440d23cd48/src/libraries/Common/src/System/Net/Security/NegotiateStreamPal.Unix.cs#L461-L465)).
.NET 7 will likely return you 401 Unauthorized and not throw, on Linux/macOS/Android ([1](https://github.com/dotnet/runtime/blob/59a1e88eaa84c1d8d34862fcfc5339440d23cd48/src/libraries/System.Net.Security/src/System/Net/Security/NegotiateAuthentication.cs#L68-L70) [2](https://github.com/dotnet/runtime/blob/59a1e88eaa84c1d8d34862fcfc5339440d23cd48/src/libraries/System.Net.Security/src/System/Net/Security/NegotiateAuthentication.cs#L289-L294)).
.NET 8 is back to throwing PlatformNotSupportedException.

Just wanted to follow-up on this bit. I tried on a domain-joined MacOS with net7.0-maccatalyst, and it works just fine there. The default credential correctly authenticates with the server using the provided repo.

**Comment by karelz** [comment]
@dotMorten would you be able to validate daily 9.0 build with the fix above?
That will help us to justify the backport to 8.0.

**Comment by dotMorten** [comment]
Sure. Bring it!

**Comment by karelz** [comment]
@dotMorten will you have time to test it out on your side please? Thanks!

**Comment by dotMorten** [comment]
@rzikm is `$DOTNET_ROOT/shared/Microsoft.NetCore.App/<version>` correct? The new file is about 3 times smaller than the one already there, and I see this DLL spread out all over the Android workloads:

<img width="608" alt="image" src="https://github.com/dotnet/runtime/assets/1378165/0528373f-eef7-4edc-8caa-4b61070c2f24">

Lastly I don't see this DLL (or any of the other DLLs) in the generated Android package.

**Comment by rzikm** [comment]
Uh... I am not that familiar with how Android packages get build. But my intuition tells me to try replace the one which has the .net 8 preview version in path (maybe all of them)

As for size, my guess is that published DLLs are distributed "ready to run", i.e. with pre-jitted machine code, which is not being done for dev builds. But I may be wrong.

@simonrozsival is what we are attempting the right way of testing private bits on Android? If not, can you please help?

**Comment by steveisok** [comment]
@rzikm @dotMorten In order to validate, you should be able to replace the assemblies in the android runtime packs that are installed via `dotnet workload install`. @dotMorten, from your image, that is all of the runtime packs (or the versions you want to replace) under `C:\Program Files\dotnet\packs`. After that, clean and rebuild and you should be using the update assembly. 

I don't think you'll see the actual assemblies in your apk because the android sdk team has a crafty resource bundling scheme.

**Comment by steveisok** [comment]
@rzikm unlike desktop, by default mobile configurations produce self contained apps. They will not use the shared framework.

**Comment by dotMorten** [comment]
Got it! I guess I didn't clean up enough the first time around. I can confirm that the fix works!

**Comment by karelz** [comment]
Fixed in main (9.0) in PR #91160 and 8.0 (RC2) in PR #91753
